File: src/ipc/events.ts
import { listen as tauriListen, type UnlistenFn } from "@tauri-apps/api/event";
import { z } from "zod";

import {
  EVENT_NAMES,
  type EventName,
  type EventPayloads,
} from "./generated/contracts";

type EventCallback<E extends EventName> = (payload: EventPayloads[E]) => void;

const refreshProgressSchema = z.object({
  status: z.string(),
});

const dataInvalidatedSchema = z.object({
  scope: z.string(),
});

const settingsChangedSchema = z.object({
  revision: z.number().int(),
});

const accountSessionChangedSchema = z.object({
  reason: z.enum([
    "login_started",
    "login_completed",
    "login_cancelled",
    "login_failed",
    "logged_out",
  ]),
});

const platformStateChangedSchema = z.object({
  kind: z.string(),
});

const updateProgressSchema = z.object({
  status: z.string(),
});

/**
 * Per-event Zod schemas. Must stay aligned with generated `EventPayloads`
 * and Rust `src-tauri/src/ipc/events.rs`.
 */
const eventSchemas: {
  [K in EventName]: z.ZodType<EventPayloads[K]>;
} = {
  [EVENT_NAMES.refreshProgress]: refreshProgressSchema,
  [EVENT_NAMES.dataInvalidated]: dataInvalidatedSchema,
  [EVENT_NAMES.settingsChanged]: settingsChangedSchema,
  [EVENT_NAMES.accountSessionChanged]: accountSessionChangedSchema,
  [EVENT_NAMES.platformStateChanged]: platformStateChangedSchema,
  [EVENT_NAMES.updateProgress]: updateProgressSchema,
};

export function parseEventPayload<E extends EventName>(
  event: E,
  payload: unknown,
): EventPayloads[E] {
  return eventSchemas[event].parse(payload);
}

export async function subscribeToEvent<E extends EventName>(
  event: E,
  callback: EventCallback<E>,
  listen: EventListenerTransport = eventListenerTransport,
): Promise<UnlistenFn> {
  return listen(event, (payload) => {
    callback(parseEventPayload(event, payload));
  });
}

export const eventListenerTransport: EventListenerTransport = (
  event,
  callback,
) =>
  tauriListen(event, ({ payload }) => {
    callback(payload);
  });

export type EventListenerTransport = (
  event: EventName,
  callback: (payload: unknown) => void,
) => Promise<UnlistenFn>;

export { EVENT_NAMES };

File: src-tauri/src/ipc/events.rs
//! Typed IPC event names and payloads.
//!
//! Events are lossy notifications. Payloads are always JSON objects (never unit/`null`).
//! The frontend re-queries authoritative command data after invalidation events.

use serde::Serialize;

/// Versioned event names (must stay aligned with `contract.rs` / generated TS).
pub(crate) mod names {
    pub(crate) const REFRESH_PROGRESS: &str = "burnly://v1/refresh-progress";
    pub(crate) const DATA_INVALIDATED: &str = "burnly://v1/data-invalidated";
    pub(crate) const SETTINGS_CHANGED: &str = "burnly://v1/settings-changed";
    pub(crate) const ACCOUNT_SESSION_CHANGED: &str = "burnly://v1/account-session-changed";
    #[expect(dead_code, reason = "reserved until platform-state emitters ship")]
    pub(crate) const PLATFORM_STATE_CHANGED: &str = "burnly://v1/platform-state-changed";
    #[expect(dead_code, reason = "reserved until update-progress emitters ship")]
    pub(crate) const UPDATE_PROGRESS: &str = "burnly://v1/update-progress";
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct RefreshProgressEvent {
    pub(crate) status: &'static str,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct DataInvalidatedEvent {
    pub(crate) scope: &'static str,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct SettingsChangedEvent {
    pub(crate) revision: i64,
}

/// Why the account session view may have changed. Frontend always re-queries.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
#[serde(rename_all = "snake_case")]
pub(crate) enum AccountSessionChangeReason {
    LoginStarted,
    LoginCompleted,
    LoginCancelled,
    LoginFailed,
    LoggedOut,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct AccountSessionChangedEvent {
    pub(crate) reason: AccountSessionChangeReason,
}

/// Reserved for future platform capability / lifecycle notifications.
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
#[expect(dead_code, reason = "wire type reserved until platform events ship")]
pub(crate) struct PlatformStateChangedEvent {
    pub(crate) kind: &'static str,
}

/// Reserved for update-progress UI (mirrors refresh-progress shape today).
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
#[expect(dead_code, reason = "wire type reserved until update progress ships")]
pub(crate) struct UpdateProgressEvent {
    pub(crate) status: &'static str,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn account_session_payload_is_object_not_null() {
        let json = serde_json::to_value(AccountSessionChangedEvent {
            reason: AccountSessionChangeReason::LoginCompleted,
        })
        .expect("serialize");
        assert!(json.is_object());
        assert_eq!(json["reason"], "login_completed");
    }

    #[test]
    fn settings_changed_payload_includes_revision() {
        let json = serde_json::to_value(SettingsChangedEvent { revision: 7 }).expect("serialize");
        assert_eq!(json["revision"], 7);
    }
}
